src/blockrelay/compactblock.cpp

```diff
@@ -0,0 +1,1216 @@
+        uint64_t nPercentileElement = static_cast<int>((nTotalEntries * 0.95) + 0.5) - 1;
+        sort(vValidationTime.begin(), vValidationTime.end());
+        nPercentile = vValidationTime[nPercentileElement];
+    }
+
+    std::ostringstream ss;
+    ss << std::fixed << std::setprecision(2);
+    ss << "Validation time (last 24hrs) AVG:" << nValidationTimeAverage << ", 95th pcntl:" << nPercentile;
+    return ss.str();
+}
+
+// Calculate the transaction re-request ratio and counter over the last 24 hours
+std::string CCompactBlockData::ReRequestedTxToString()
+{
+    LOCK(cs_compactblockstats);
+
+    double nReRequestRate = compute24hInboundRerequestTxPercentInternal();
+
+    // NOTE: Potential gotcha, compute24hInboundRerequestTxPercentInternal has a side-effect of calling
+    //       expireStats which modifies the contents of mapCompactBlocksInBoundReRequestedTx
+    // We currently rely on this side-effect for the string produced below
+    std::ostringstream ss;
+    ss << std::fixed << std::setprecision(1);
+    ss << "Tx re-request rate (last 24hrs): " << nReRequestRate
+       << "% Total re-requests:" << mapCompactBlocksInBoundReRequestedTx.size();
+    return ss.str();
+}
+
+std::string CCompactBlockData::MempoolLimiterBytesSavedToString()
+{
+    LOCK(cs_compactblockstats);
+    double size = (double)nMempoolLimiterBytesSaved();
+    std::ostringstream ss;
+    ss << "CompactBlock mempool limiting has saved " << formatInfoUnit(size) << " of bandwidth";
+    return ss.str();
+}
+
+// Calculate the average compact block size
+std::string CCompactBlockData::CompactBlockToString()
+{
+    LOCK(cs_compactblockstats);
+    double avgCompactBlockSize = average(mapCompactBlock);
+    std::ostringstream ss;
+    ss << "CompactBlock size (last 24hrs) AVG: " << formatInfoUnit(avgCompactBlockSize);
+    return ss.str();
+}
+
+// Calculate the average size of all full txs sent with block
+std::string CCompactBlockData::FullTxToString()
+{
+    LOCK(cs_compactblockstats);
+    double avgFullTxSize = average(mapFullTx);
+    std::ostringstream ss;
+    ss << "compactblock full transactions size (last 24hrs) AVG: " << formatInfoUnit(avgFullTxSize);
+    return ss.str();
+}
+
+// After a compactblock is finished processing or if for some reason we have to pre-empt the rebuilding
+// of a compactblock then we clear out the compactblock data which can be substantial.
+void CCompactBlockData::ClearCompactBlockData(CNode *pnode)
+{
+    // Remove bytes from counter
+    compactdata.DeleteCompactBlockBytes(pnode->nLocalCompactBlockBytes, pnode);
+    pnode->nLocalCompactBlockBytes = 0;
+
+    // Clear out compactblock data we no longer need
+    pnode->compactBlockWaitingForTxns = -1;
+    pnode->compactBlock.SetNull();
+    pnode->vCompactBlockHashes.clear();
+    pnode->vShortCompactBlockHashes.clear();
+    pnode->mapMissingTx.clear();
+
+    LOG(CMPCT, "Total in memory compactblockbytes size after clearing a compactblock is %ld bytes\n",
```

Comment:
missing space between compactblock and size?

Reply:
done